internal/transport/http2_client.go

```diff
@@ -1010,6 +1010,18 @@ func (t *http2Client) Close(err error) {
 	}
 	t.mu.Unlock()
 
+	// Append info about previous goaways if there were any, since this may be important
+	// for understanding the root cause for this connection to be closed.
+	_, goAwayDebugMessage := t.GetGoAwayReason()
+
```

Comment:
The current implementation of `http2Client.outgoingGoAwayHandler` holds `t.mu` when calling `WriteGoAway` on the underlying framer, i.e. it is performing I/O when holding the lock. Is this the correct thing to do?

Can/should this be replaced with the following:
```golang
// OutgoingGoAwayHandler writes a GOAWAY to the connection.  Always returns
// (false, err) as we want the GoAway to be the last frame loopy writes to the
// transport.
func (t *http2Client) outgoingGoAwayHandler(g *goAway) (bool, error) {
	t.mu.Lock()
	nextID := t.nextID - 2
	t.mu.Unlock()
	if err := t.framer.fr.WriteGoAway(nextID, http2.ErrCodeNo, g.debugData); err != nil {
		return false, err
	}
	return false, g.closeConn
}
```

Reply:
Yeah that'd be a useful optimization. We could define `maxStreamID` as `t.nextID - 2` and use it to write goAway frame as `t.framer.fr.WriteGoAway(maxStreamID, http2.ErrCodeNo, g.debugData)`.

Reply:
The call to `t.mu.Unlock()`, currently on line 1012, needs to be moved to be after this line because `t.goAwayDebugMessage` is guarded by that mutex.

Reply:
> Yeah that'd be a useful optimization and hence minimize acquiring the mutex

I don't think this is a useful optimization. This is the root of the deadlock. I don't think we need to move code around in `http2Client.Close` at all if we don't hold the lock in `outgoingGoAwayHandler` when performing the actual write. Can you please confirm that?

Reply:
Interesting, yeah but we would need to release `t.mu` before we sent out goAway frame over to cbuf, bcz `outgoingGoAwayHandler` also needs to acquire `t.mu` to get the `maxStreamID` and we are waiting for t.writerDone with the timer. We can do something like


    goAwayDebugMessage := t.goAwayDebugMessage
	t.mu.Unlock()

	// Per HTTP/2 spec, a GOAWAY frame must be sent before closing the
	// connection. See https://httpwg.org/specs/rfc7540.html#GOAWAY. It
	// also waits for loopyWriter to be closed with a timer to avoid the
	// long blocking in case the connection is blackholed, i.e. TCP is
	// just stuck.
	t.controlBuf.put(&goAway{code: http2.ErrCodeNo, debugData: []byte("client transport shutdown"), closeConn: err})
	timer := time.NewTimer(goAwayLoopyWriterTimeout)
	defer timer.Stop()
	select {
	case <-t.writerDone: // success
	case <-timer.C:
		t.logger.Infof("Failed to write a GOAWAY frame as part of connection close after %s. Giving up and closing the transport.", goAwayLoopyWriterTimeout)
	}
	t.cancel()
	t.conn.Close()
	channelz.RemoveEntry(t.channelz.ID)
	var st *status.Status
	if len(goAwayDebugMessage) > 0 {
		st = status.Newf(codes.Unavailable, "closing transport due to: %v, received prior goaway: %v", err, goAwayDebugMessage)
		err = st.Err()
	} else {
		st = status.New(codes.Unavailable, err.Error())
	}

WDYT?

Reply:
Yes, that was my suggestion.